src/NuGet.Core/NuGet.Configuration/PackageNamespaces/SearchNode.cs

```diff
@@ -0,0 +1,25 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace NuGet.Configuration
+{
+    internal class SearchNode
+    {
+        public readonly SearchNode Parent;
```

Comment:
I think we can avoid maintaining `parent` node. Let's say if the namespaces are configured in the following way.

```
  <packageNamespaces>
    <packageSource key="source1">
      <namespace id="NuGet*" />
    </packageSource>
    <packageSource key="source2">
      <namespace id="NuGet.Common" />
    </packageSource>
  </packageNamespaces>
```

I would image the Trie should look like below.

![image](https://user-images.githubusercontent.com/52756182/120369561-4a41b500-c2c8-11eb-881f-49ee021651e9.png)

If the project has a dependency on `NuGet.Protocol` package, then when searching for the prefix, if we can capture `source1` along the way then we don't need to have a parent node.

I see a Trie implementation in Visual Studio magazine https://visualstudiomagazine.com/articles/2015/10/20/text-pattern-search-trie-class-net.aspx which has `Parent` node but I guess we can remove it if we don't need it. Please let me know if I missed any scenarios.

Reply:
> I think we can avoid maintaining `parent` node. Let's say if the namespaces are configured in the following way.
> 
> ```
>   <packageNamespaces>
>     <packageSource key="source1">
>       <namespace id="NuGet*" />
>     </packageSource>
>     <packageSource key="source2">
>       <namespace id="NuGet.Common" />
>     </packageSource>
>   </packageNamespaces>
> ```
> 
> I would image the Trie should look like below.
> 
> ![image](https://user-images.githubusercontent.com/52756182/120369561-4a41b500-c2c8-11eb-881f-49ee021651e9.png)
> 
> If the project has a dependency on `NuGet.Protocol` package, then when searching for the prefix, if we can capture `source1` along the way then we don't need to have a parent node.
> 
> I see a Trie implementation in Visual Studio magazine https://visualstudiomagazine.com/articles/2015/10/20/text-pattern-search-trie-class-net.aspx which has `Parent` node but I guess we can remove it if we don't need it. Please let me know if I missed any scenarios.

Parent is used for retracing back. 
Let see imagine we have 3 Prefixes from different sources:
source1: Hello*
source2: Hel.*
source3:HelloWz.*
Package: HelloWorld

In this case: Above algorithm we'll stuck on source3 with 'HelloW' because it's longest matching prefix at the moment, but it start diverge after that so source3 is not good anymore. We need to find source1 which is better candidate now. 
That falling back to just Hello need parent. Please let me know if you have any other question.

Reply:
Correct. I agree but if we maintain the hits along the way, we can eliminate the `parent` node. In the above example you have mentioned, we will reach "HelloW" node only after reaching "Hello" node. Hence if we capture the source at "Hello" node itself then we don't to walk up the Trie data structure when there is no char matching after "HelloW" node.

Reply:
If we stop at Hello what happens if there'is package called `HellWz.1` do exists? We can't use prefix on this package now.

Reply:
Let's say if we are parsing `HelloWz.1` package, during traversal we first capture `source1` as it matches `Hello.*` later when we find the longest prefix match at `HelloWz.*` then we will overwrite the result with `source3`.

Reply:
Result ~ tree?
Source1, Source2, Source3 all sharing same trie tree, nothing we'll be overwritten, this tree created only once and used again and again. All context are saved into 1 trie tree.